cpp/src/mlt/decode/int_template.hpp

```diff
@@ -0,0 +1,226 @@
+#pragma once
+
+#include <mlt/decode/int.hpp>
+
+#include <cassert>
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+namespace mlt::decoder {
+
+#pragma region Public
+
+template <typename T, typename TTarget, bool isSigned>
+    requires((std::is_integral_v<T> || std::is_enum_v<T>) && (std::is_integral_v<TTarget> || std::is_enum_v<TTarget>) &&
+             sizeof(T) <= sizeof(TTarget))
+void IntegerDecoder::decodeIntArray(const std::vector<T>& values,
+                                    std::vector<TTarget>& out,
+                                    const StreamMetadata& streamMetadata) noexcept(false) {
+    using namespace metadata::stream;
+    using namespace util::decoding;
+    switch (streamMetadata.getLogicalLevelTechnique1()) {
+        case LogicalLevelTechnique::NONE: {
+            out.resize(values.size());
+            const auto f = isSigned ? [](T x) noexcept { return static_cast<TTarget>(decodeZigZag(x)); } : [](T x) noexcept { return static_cast<TTarget>(x); };
+            std::ranges::transform(values, out.begin(), f);
+            return;
+        }
+        case LogicalLevelTechnique::DELTA:
+            if (streamMetadata.getLogicalLevelTechnique2() == LogicalLevelTechnique::RLE) {
+                if (streamMetadata.getMetadataType() != metadata::stream::LogicalLevelTechnique::RLE) {
+                    throw std::runtime_error("invalid RLE metadata");
+                }
+                const auto& rleMetadata = static_cast<const RleEncodedStreamMetadata&>(streamMetadata);
+                out.resize(rleMetadata.getNumRleValues());
+                rle::decodeInt<T, TTarget>(values, out, rleMetadata.getRuns());
+                decodeZigZagDelta(out, out);
+            } else {
+                out.resize(values.size());
+                decodeZigZagDelta(values, out);
+            }
+            break;
+        case LogicalLevelTechnique::COMPONENTWISE_DELTA:
+            if constexpr (std::is_same_v<TTarget, std::int32_t> || std::is_same_v<TTarget, std::uint32_t>) {
+                out.resize(values.size());
+                std::ranges::transform(values, out.begin(), [](auto x) { return static_cast<TTarget>(x); });
+                vectorized::decodeComponentwiseDeltaVec2(out);
+                break;
+            }
+            throw std::runtime_error("Logical level technique COMPONENTWISE_DELTA not implemented for 64-bit values");
```

Comment:
I think this can be a `static_assert` instead.

Reply:
That condition is only required for that branch in the switch, while the `static_assert` fails for all invocations.  Perhaps if I moved the switch case to a separate method, but the problem will go away when that functionality is implemented.